Add Polyphonic Chord Training Support for Musical Deep Learning

cucuwritescode opened the pull request:
this PR addresses the limitation of single-note synthesis by providing:
  -harmonic relationships between notes within different FM timbres
  -realistic musical articulations 
  -polyphonic support for complex harmonic configurations

cucuwritescode committed:
chord/voicing dictionary

in progress, helper function to generate inversions with bass note for different chord types in an automatic manner. also a dictionary with multiple voicing configurations, from simple triads to complex extended chords

cucuwritescode committed:
Create .DS_Store

fcaspe commented on dx7pytorch/dataset.py:
```diff
@@ -4,6 +4,87 @@
+        high_notes = intervals[inversion_num:]
+
+        #take the first note and shift it an octave up
+        low_notes= [note+12 for note in intervals[:inversion_num]]
```
I think this is a great idea, no need to shift them down, because we might as well have a lower root and that's it.

fcaspe commented:
This is looking good! Great work so far. 

I would continue now with these procedures:

1. Add this programmatic chord generation to the dataset constructor. Identify clearly each function.
2. Pick a random voicing and synthesize a chord, instead of synthesizing a single note. Use a chance of 50% of picking between synthesizing a single note or synthesizing a random chord voicing. Use the random picked velocity of the instance **for all notes (for now)** of the synthesized voicing.

Let's see how does that sound before proceding with more additions to the chord/arpeggio generator!!

Thanks so much,
Fran

cucuwritescode committed:
add polyphonic chord synthesis support

-extend to dx7pytorch to generate harmonic chord content alongside single notes for richer musical training data
-implements chord voicing system with automatic inversion generation
-added 50/50 randomised chord/single note sampling for balanced harmonic/single tone training
-also added a simple midi to note conversion algorithm and prints on the generated plots with chord type and root note info

cucuwritescode commented:
only downside is im using a mixing approach instead of “true polyphony”. individual notes are synthesised separately and mixed together rather than using the DX7 engine’s native poly capabilities. I couldn’t get it to work tonight, maybe you see an easy fix? 
or we could keep the mixing approach anyway since it seems to work fine, though not sure if it’s the best possible way of doing it

fcaspe commented:
I think this is the right approach, so as not to introduce modifications in the C++ code. The only thing, you should remove the generate_chord function from the synth module, and create a generate chord function in the dataset class. The synth class is a mirror of the C++ component, so there is no chord generation function. It's all independent voices generated concurrently and summed together.

Other than that it is looking good. Let's move the chord generation to the dataset class and start thinking how can we introduce variation to the chord generation. One possibility could be to draw velocity values for each chord note from a.normal distribution centered around the main velocity value.

Then we could look at arpeggios, but let's cintinue with intra-chord velocity variations.

cucuwritescode committed:
moved chord generation to the dataset class

fcaspe commented:
This is sounding so well!! Great work so far with the chord generation!
Looking great so far. What I would be missing here is this:

(1) Render arpeggios as an alternative of a chord, using the same exact voicings.
this way you could set a probability for 1/3 single note, 1/3 chord 1/3 arpeggio.

(2) Instead of using a single velocity for all the notes of the chords or arpeggios, model a Gaussian distribution with a mean of the given velocity, and a standard deviation of let's say 20. So you draw for each note of the chord/arpeggio a new velocity value that you use to synthesize. Clip the values below 1 to 1 and above 127 to 127.

What do you think? Looking forward to see what comes out of this. Thanks so much!
Fran

cucuwritescode committed:
add arpeggios and velocity variation

cucuwritescode commented:
Implemented both requested features:
arpeggios: added as third synthesis option (1/3 probability each for single/chord/arpeggio). notes are placed sequentially across the sample duration

gaussian velocity variation: each note in chords/arpeggios now draws velocity from N(base_velocity, 20), clipped to [1,127] for more human-like dynamics

also updated the audition notebook for testing

fcaspe commented:
Hey Facu, this looks great, thanks so much for adding the arpeggios, they sound very nice. I will merge this into the refactor branch. 
I will then add a few mods such as staccato and legato for arpeggios and merge into main.

fcaspe commented:
Just added a few corrections and pushed to `master`. Thanks so much again, and hope you have happy holidays!
